### DISCUSSION TITLE
Console filter regular expression, option to save across UGS invocations

### DISCUSSION THREAD

**Discussion by bLouChip** [winder/Universal-G-Code-Sender#2459] [Ideas]
I noticed in Controller Options that a saved regular expression is applied to  Comment Remover.  Two questions here:
1) this option is greyed out (locked) and checked (enabled), can it be unlocked so that it can be disabled or enabled as desired ? 
2) can another option be added which is the saved filter as a regular expression for the Console streaming ? 

Here is a regexp that Breiler assisted me in writing.  I find it very useful for the Console streaming.
regexp will filter out "ok **" and jogging cmds and ">>> white space" from the Console streaming log.
`^(?!ok\s*$)(?!>>> \$J=.*$)(?!>>>\s*$)`

**Comment by breiler** [comment]
> this option is greyed out (locked) and checked (enabled), can it be unlocked so that it can be disabled or enabled as desired ?

No, the GRBL protocol does not support comments which would otherwise cause the controller to error out, it is therefore not optional and enabled by default.

> can another option be added which is the saved filter as a regular expression for the Console streaming ?

I had the ambition to add some new features to the console such as colors (#2043) and folding (group multiple lines into one). A feature for filtering might be a good idea as well to add to that list.

**Comment by bLouChip** [comment]
re. grbl comments-  then perhaps I'm confusing the the linkage of Comment Remover option with grbl comments no longer being seen in the Console log stream; in previous UGS releases (going back 6 months or more), I recall seeing grbl comments in the Console log stream, that was useful to me.  As noted with the Console log filter field now available (right click in log), grbl comments can be filtered there if undesired.  So are you saying that previously those grbl comments were visible in the Console log but not actually sent to grbl ?  Or they were being sent and that caused some problems ? 

re. adding the Console filter regexp as a saved option, thank you.

**Reply by breiler** [reply]
> I recall seeing grbl comments in the Console log stream, that was useful to me.

Yeah, I think that I removed that feature with the intention to replace it with another method of logging. I added the issue #2461 so I don't forget about it. But I have not yet decided which priority it has yet.

> So are you saying that previously those grbl comments were visible in the Console log but not actually sent to grbl ?

Exactly

**Comment by bLouChip** [comment]
@breiler ,   re. OP #2 request...

> can another option be added which is the saved filter as a regular expression for the Console streaming ?

perhaps some confusion here, what I meant to suggest _is a **UGS launch/exit cycle persistent** save of the regexp filter string field_ that already exists for the console log.  I was just suggesting to implement a list of filter strings as a Controller Option.  As it is now, the strings are only persistent during the current invocation of UGS in this filter dialogue. 

I use Console Log filters frequently to analyze the job during runtime as well as post runtime.  Extremely useful. 

I save my Console Log filter streams now in the Controller Options window as disable regexp strings for gcode, only because it's more convenient that putting them in an external text file. 
Thanks for considering.

![Screenshot 2024-03-23 092519](https://github.com/winder/Universal-G-Code-Sender/assets/88944660/94376395-0424-4440-a4d9-baecdc3529a4)

![Screenshot 2024-03-23 092611](https://github.com/winder/Universal-G-Code-Sender/assets/88944660/0e92be6d-f1b2-4c07-b409-0fcdf6f16e11)

![image](https://github.com/winder/Universal-G-Code-Sender/assets/88944660/1b3eb758-da4c-4f80-a964-4c35288f7c8c)

**Comment by bLouChip** [comment]
@breiler -  just checking if you're considering adding an option to save the Console filter expression ? Its very handy.  It could also implement the filtering of gcode comments for those that don't want them.  I prefer gcode comments.  But either way, everyone is happy :) 
I use Console filtering in two use cases...
1) I filter OUT the pesky jog commands, 
`^(?!ok\s*$)(?!>>> \$J=.*$)(?!>>>\s*$)`

2) I filter IN all but [PRB:* msgs in context with gcode probe commands to monitor THC Z error during plasma cutting- 
`^\[PRB:.*$|^>>> G38\.(2|3).*$|.*MegaV.*$|^>>> G00G17G40.*$|\$HZ.*$`

**Reply by breiler** [reply]
It is a good idea, but is not on top of my priority list at the moment.
